source("./update_taxonomy_refseq.R")

# TESTS for function is_taxonomy_incomplete
### Function for determining if original greengenes taxonomy is incomplete.
# Incomplete is not assigned at specified taxonomic level.
# Returns boolean TRUE if taxonomy is incomplete.

current_string <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lactobacillus;s__buchneri"
is_taxonomy_incomplete(current_string, "spcs")
is_taxonomy_incomplete(current_string, "gen")
is_taxonomy_incomplete(current_string, "fam")

current_string2 <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lactobacillus;s__"
is_taxonomy_incomplete(current_string2, "spcs")
is_taxonomy_incomplete(current_string2, "gen")
is_taxonomy_incomplete(current_string2, "fam")

current_string3 <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__;s__"
is_taxonomy_incomplete(current_string3, "spcs")
is_taxonomy_incomplete(current_string3, "gen")
is_taxonomy_incomplete(current_string3, "fam")

current_string4 <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae"
is_taxonomy_incomplete(current_string4, "spcs")
is_taxonomy_incomplete(current_string4, "gen")
is_taxonomy_incomplete(current_string4, "fam")

current_string5 <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;"
current_string5 <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__;g__;s__"
current_string5 <- "k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales"
is_taxonomy_incomplete(current_string5, "spcs")
is_taxonomy_incomplete(current_string5, "gen")
is_taxonomy_incomplete(current_string5, "fam")





# Reading files for next tests

# fasta file
data_fasta <- readDNAStringSet("./data_for_tests/bacteria_dna_sequences.fasta")

# Reading microbial data base for tests
microbial_database <- blast(db = "./data_for_tests/16SMicrobialDB/16SMicrobial")

# set NCBI's entrez api key
Sys.setenv(ENTREZ_KEY = "ed4870836e8f61529227d9176a7c4a994c07")
# Check that key variable is in path.
getkey(service = "entrez")
# Check if blast is in PATH.
Sys.which("blastn")

head(data_fasta)

# Converting fasta file to text.
#as.character(fasta1[[1]])
# Searching fasta file by otu_id
#data_fasta[data_fasta@ranges@NAMES == "4fdb872697ff4712d1408c2a31c881ef"]


# TESTS blast_n_get_ncbi_tax.
### Function for blasting sequences against microbial refseq 16s database and returning the taxonomy from NCBI's taxonomy database.
# Orders results by bits and percent of identity. Then chooses the first match with identity percent.
# Returns vector of size 3 including:
# 1) bool if % indentity is at least given number, def. 97 and bits >= 100;
# 2) float of blast result's % identity;
# 3) taxonomy in ncbi's format.

fasta1 <- data_fasta[1,]
fasta1
fasta2 <- data_fasta[10,]
fasta2
fasta3 <- data_fasta[30,]
fasta3

ref_seq_taxonomy1 <- blast_n_get_ncbi_tax(fasta1, microbial_database = microbial_database)
ref_seq_taxonomy1

ref_seq_taxonomy2 <- blast_n_get_ncbi_tax(fasta2, microbial_database = microbial_database)
ref_seq_taxonomy2

ref_seq_taxonomy3 <- blast_n_get_ncbi_tax(fasta3, microbial_database = microbial_database)
ref_seq_taxonomy3




# TESTS parse_ncbi_to_gg
### Function for parsing the NCBI's taxonomy into greengenes format.
# Input is NCBI's taxonomy as a nested list containing information about each level of the classification.
# Returns string of taxonomy in greengenes format.
# This test uses the output generated by the previous test (blast_n_get_ncbi_tax).

parse_ncbi_to_gg(ref_seq_taxonomy1[4])
parse_ncbi_to_gg(ref_seq_taxonomy2[4])
parse_ncbi_to_gg(ref_seq_taxonomy3[4])


# TESTS update_taxonomy_refseq
### Function for updating a taxonomy table generated by QIIME 2 (taxonomy.tsv).
#The updated table can be joined to the corresponding BIOM file to perform the rest of the analyses of the QIIME2 pipeline.
# Inputs are:
#   1) taxonomy_table: a taxonomy table in tsv format is dataframe type.
#   2) data_fasta: fasta file for all the otus in taxonomy table (with the same otu_ids).
#   3) level: string describing the desired level of analysis deafault:"spcs"; also: "gen" or "fam".
#   4) phyl_group: string describing the analysed group. deafault: "bacteria"; also: "fungi".
#   5) update_all: bool. Indicates if the user wants to reassign all the otus or only the otus with incomplete taxonomies. Default: FALSE
# Returns a dataframe with the updated taxonomy using the same layout as the original input.
# Dataframe can be writen to tsv format to use with QIIME2.

# Reading required data.

# Load taxonomy table to be updated with NCBI's taxonomy.
#tax_table <- read.table("C:/Users/marce/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/chelo/hidro_agave_diversidad/2_resultados/9_tsv_gg_sk/taxonomy.tsv", sep = "\t", header = TRUE)
bacteria_tax_table <- read.table("./data_for_tests/bacteria_taxonomy.tsv", sep = "\t", header = TRUE)
bacteria_tax_table

# Load sequences to be analyzed in fasta format. Ids corresponding to taxonomy OTUs.
bacteria_data_fasta <- readDNAStringSet("./data_for_tests/bacteria_dna_sequences.fasta")
# Check fasta sequences
bacteria_data_fasta

# Reading microbial data base for tests
microbial_database <- blast(db = "./data_for_tests/16SMicrobialDB/16SMicrobial")
microbial_database

# set NCBI's entrez api key
Sys.setenv(ENTREZ_KEY = "ed4870836e8f61529227d9176a7c4a994c07")
# Check that key variable is in path.
getkey(service = "entrez")
# Check if blast is in PATH.
Sys.which("blastn")


# This is the copy of taxonomy file that we are modifying.
head(bacteria_tax_table)
bacteria_tax_table$Taxon <- as.character(bacteria_tax_table$Taxon)
head(bacteria_tax_table)
bacteria_tax_table <- as.data.frame(apply(bacteria_tax_table, 2, as.character), stringsAsFactors = FALSE)
head(bacteria_tax_table)

select(bacteria_tax_table, Taxon)

bacteria_tax_table[1,]

# este si sirve
(bacteria_tax_table %>% filter(Feature.ID == "4fdb872697ff4712d1408c2a31c881ef") %>% select(Taxon))[1, 1]

select(bacteria_tax_table[1,], Taxon)[1, 1]

nrow(bacteria_tax_table)

typeof(bacteria_tax_table)

tax_table_updated <- update_taxonomy_refseq(taxonomy_table = bacteria_tax_table, data_fasta = bacteria_data_fasta, level = "spcs", phyl_group = "bacteria", update_all = FALSE)

head(tax_table_updated)

write.table(new_tax3, file = "./taxonomy_updated_picrust_13_8.tsv", sep = "\t", row.names = FALSE)


